Isara UI fixes

meguiraun opened the pull request:
hi,

The samples view page was not working correcly for our Isara. I believe the main reason was the cell logic, so instead of "faking" cells for our case I created the new method getSampleTableSingleCell, a simplfied version of getSampleTable but with less iterations for non existing cells, the logic was started to be very obfuscated.

I could not test against FlexHCD... how do you test for it?

meguiraun committed:
reset filter in samplew view mode change

meguiraun committed:
hacking our sample list view

meguiraun committed:
initisalize to false

meguiraun committed:
fix isara view to show pucks above 16

meguiraun committed:
Fixes for isara like view

meguiraun committed:
do not render if no sample to display

meguiraun committed:
do not break the page, just alert

meguiraun committed:
bit of linting cleanup

meguiraun committed:
no console

jbflo commented:
Hi @meguiraun , 

Maybe we can take some time to work on this together, because this PR break a bit the Table view  for FLEXHCD.
getSampleTableSingleCell is what you'll use for ISARA 
so getSampleTable should remain the same right ?

meguiraun committed:
undo changes in getSampleTable

meguiraun commented:
> Hi @meguiraun ,
> 
> Maybe we can take some time to work on this together, because this PR break a bit the Table view for FLEXHCD. getSampleTableSingleCell is what you'll use for ISARA so getSampleTable should remain the same right ?

correct! I just fixed the _getSampleTable_ so it stays as before

jbflo commented on ui/src/components/SampleGrid/SampleGridTable.css:
```diff
@@ -17,6 +21,7 @@
 
 .samples-grid-table-card-body {
   padding: 1.5em 0.5em 0.5em 0.5em;
+  overflow: scroll;
```
I think users prefer scrolling vertical than horizontal.
I would suggest to remove this style  overflow: scroll; ?

jbflo commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -528,8 +530,8 @@ class SampleGridTableContainer extends React.Component {
-
-      puck_code = allPuckSample[0].containerCode || '';
+      puck_code =
+        allPuckSample.length > 0 ? allPuckSample[0].containerCode : '';
```
in this logic puck_code return undefined when there is no containerCode

to avoid displaying undefined
below maybe you can add  if (puck_code !== ''  && puck_code !== undefined) {

jbflo commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -1194,10 +1386,11 @@ class SampleGridTableContainer extends React.Component {
-              {this.getManualSamples()}
-              {this.getSampleTable('auto')}
-            </Col>
+            {this.getManualSamples()}
```
{this.getManualSamples()}
should be in a separate Row

Or like this
`<Col className="col-sm-12 mb-1"> {this.getManualSamples()} </Col>`

jbflo replied:
![image](https://github.com/mxcube/mxcubeweb/assets/7275231/17b69503-f355-4813-a781-b13760e7bd6e)

jbflo commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -1212,7 +1405,9 @@ class SampleGridTableContainer extends React.Component {
           >
             <div className="selection-rubber-band" id="selectionRubberBand" />
             {this.getManualSamples()}
```
{this.getManualSamples()}
should be in a separate ROW

Or  like this
`<Col className="col-sm-12 mb-1">
      {this.getManualSamples()}
    </Col>`

jbflo commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -965,6 +1047,117 @@ class SampleGridTableContainer extends React.Component {
+              marginRight: '2px',
+            }}
+          >
+            {this.itemsControls(Number(puck.name), null)}
```
{this.itemsControls(1, Number(puck.name))}

this will work

jbflo replied:
[Screencast from 09-27-2023 04:15:34 PM.webm](https://github.com/mxcube/mxcubeweb/assets/7275231/bd662d5f-0996-4afa-b679-7ccfbd4afdb8)

jbflo commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -965,6 +1047,117 @@ class SampleGridTableContainer extends React.Component {
+              this.displayPuckCellContextMenu(
+                e,
+                puckMenuID,
+                Number(puck.name),
```
```
 onClick={(e) => {
              this.displayPuckCellContextMenu(
                e,
                puckMenuID,
                1,
                Number(puck.name),
              );
            }}
```

jbflo replied:
[Screencast from 09-27-2023 04:15:34 PM.webm](https://github.com/mxcube/mxcubeweb/assets/7275231/d11c718c-bd95-4e56-a075-b5810cda7ed3)

meguiraun commented:
I noticed the sample filtering has stopped working for some change in this branch, setting as WIP while I fix it

edit: fixed!

meguiraun committed:
fixing filtering

jbflo commented on ui/src/components/SampleGrid/SampleGridTable.css:
```diff
@@ -6,6 +6,10 @@
 }
 
+.samples-grid-table {
+  flex-wrap: nowrap;
```
I think users prefer scrolling vertical than horizontal.
I would suggest to remove this style flex-wrap: nowrap; ?

jbflo commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -965,6 +1053,135 @@ class SampleGridTableContainer extends React.Component {
+            colsmP = colsm;
+          }
+          tableCell.push(
+            <Col sm={colsmP} key={`puck-${puck.name}`}>
```
suggest adding className="mt-2" in Col

axelboc commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -380,6 +382,9 @@ class SampleGridTableContainer extends React.Component {
         hasLimsData,
       );
+      if (this.props.filterOptions.puckFilter != '') {
+        fi &= sample.puck_no == this.props.filterOptions.puckFilter;
```
Can this be changed to a strict equality, `===`?

axelboc commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -758,15 +766,92 @@ class SampleGridTableContainer extends React.Component {
+      return null;
+    }
+    // const key = sample.sampleID;
+    const key = puck + ':' + sampleNumber;
```
If you move this to the top of the function, you can use it inside the `find` loop.

axelboc commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -758,15 +766,92 @@ class SampleGridTableContainer extends React.Component {
+    const sample = Object.values(this.props.sampleList).find(
+      (sample) => sample.location === puck + ':' + sampleNumber,
+    );
+    if (typeof sample === 'undefined') {
```
You can just write `sample === undefined` here.

axelboc commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -758,15 +766,92 @@ class SampleGridTableContainer extends React.Component {
     return sampleItemList;
   }
 
+  getPuckSampleItems(puck, sampleNumber) {
```
This function returns JSX, so it should be prefixed with `render`.

axelboc commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -758,15 +766,92 @@ class SampleGridTableContainer extends React.Component {
+      'samples-grid-table-item-collected': isCollected(sample),
+    });
+
+    const settings = {
+      dots: false,
+      infinite: false,
+      speed: 100,
+      slidesToShow: 6,
+      slidesToScroll: 6,
+    };
```
This looks like a static object, so you can move it to the top of the file (and use constant case: `const SETTINGS = ...`) to avoid recreating it every time.

axelboc commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -758,15 +766,92 @@ class SampleGridTableContainer extends React.Component {
+      slidesToScroll: 6,
+    };
+
+    let contextMenuID = 'samples-grid-table-context-menu';
+    if (this.currentSample(sample.sampleID)) {
+      contextMenuID = 'samples-grid-table-context-menu-mounted';
+    }
```
To avoid `let`:

```js
const contextMenuID = `samples-grid-table-context-menu${this.currentSample(sample.sampleID) ? '-mounted' : ''}`
```

axelboc commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -810,14 +895,17 @@ class SampleGridTableContainer extends React.Component {
     const scContent = this.props.sampleChanger.contents;
     const tableCell = [];
 
+    if (typeof scContent.children === 'undefined') {
```
`scContent.children === undefined`

axelboc commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -965,6 +1053,135 @@ class SampleGridTableContainer extends React.Component {
+    return null;
+  }
+
+  getSampleTableSingleCell(colsm) {
```
`renderSampleTableSingleCell`

axelboc commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -965,6 +1053,135 @@ class SampleGridTableContainer extends React.Component {
+    const scContent = this.props.sampleChanger.contents;
+    const tableCell = [];
+
+    // Do not try to render is SC is empty
```
`if SC is empty` (same below, though it might be worth merging the two `if`s)

axelboc commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -965,6 +1053,135 @@ class SampleGridTableContainer extends React.Component {
+        (currentPuckFilter == '' || currentPuckFilter == puck.name) &&
+        filterList[0].length > 0
+      ) {
+        const nbpuck = [];
+        nbpuck.push(puck);
```
I'm struggling to see how `nbpuck` can contain more than one element (with regards to the `if` blocks below).

axelboc commented on ui/src/containers/SampleGridTableContainer.jsx:
```diff
@@ -965,6 +1053,135 @@ class SampleGridTableContainer extends React.Component {
+      return null;
+    }
+
+    scContent.children.map((puck) => {
```
I think this could be written as:

```
return scContent.children.map(...);
```

... instead of initialising and pushing to the `tableCell` array.

jbflo commented:
Hi @meguiraun 
Can I push some changes to this branch ?
so i don't need to create another  PR

meguiraun commented:
thank you guys for your comments! I will not be able to adress them in the next couple of days :(

Yeah, of course, @jbflo  push to this branch as much as you want ;)

jbflo committed:
We should not log/print queue json this make request slow

jbflo committed:
FIx unique key warning

jbflo committed:
adjust css for sample list view

jbflo committed:
Update sample list view

Fix filter  by cell and puck logic ,
specify for Single Cell Table View
move function from child to Parent
Add logic to control all Samples Menu

jbflo committed:
LINT / Run Format

jbflo committed:
function returns JSX,  named should be  prefixed with render

jbflo commented:
That should be it, 
the Files are a bit huge in therm of line of code but it is for handling both SC content with multiple Cell or not

![image](https://github.com/mxcube/mxcubeweb/assets/7275231/c973011b-32a7-4697-b325-d2737d94af4a)
![image](https://github.com/mxcube/mxcubeweb/assets/7275231/2a26c891-cb7d-4185-8dfc-2215b2dfe968)

Commit:
Rebase to Develop

LInt And replace render function  by a js function

jbflo commented:
Hey @meguiraun 
Just to let you know i was done with the changes.  we may probably review/merge it before it got Conflict.

meguiraun commented:
hej @jbflo!

I did not have much chance of testing it, neither showing to our scientist, but yeah, better merge now before is late. We can always tweak things later

meguiraun commented:
and thanks for all the job :)!

meguiraun commented:
I cannot approve this myself since I am the MR creator, but I fully approve it.

Commit:
Small adjust for small screen

marcus-oscarsson commented on mxcube3/core/components/queue.py:
```diff
@@ -2201,7 +2201,7 @@ def queue_enable_item(self, qid_list, enabled):
             self.set_enabled_entry(qid, enabled)
 
-        logging.getLogger("MX3.HWR").info("[QUEUE] is:\n%s " % self.queue_to_json())
+        # logging.getLogger("MX3.HWR").info("[QUEUE] is:\n%s " % self.queue_to_json())
```
You can perhaps remove line or put level at debug ?

jbflo replied:
I remove it, cause it take time to print/log this

marcus-oscarsson commented on ui/src/containers/SampleListViewContainer.jsx:
```diff
@@ -471,6 +575,21 @@ class SampleListViewContainer extends React.Component {
+      this.filter(key),
+    );
+
+    debugger;
```
Watch out you have a debugger statement on the loose :)

jbflo commented:
Hey @marcus-oscarsson
We are a bit done with this, for now , you can review/merge it once you got the time. 

next step is to simplifying the SampleGridTableContainer and convert it to function component
for another PR.

marcus-oscarsson commented:
Ok, made a quick review I spotted a few things. Great initiative to make it a functional component !

Commit:
Remove commented line

Commit:
Eslint / lint

marcus-oscarsson commented:
Hey guys, are you still working on this or can I do a final pass and merge ?

meguiraun commented:
although I created myself I guess is now @jbflo the main developer 😜... looks good to me for merging

jbflo commented:
@marcus-oscarsson , We done thanks
